### PR TITLE
Dictionary - add missing data types

### PR DESCRIPTION
**Pull request by edualonso**
Added missing supported types to RealmMap:
- STRING
- BINARY
- BINARY_NON_PRIMITIVE
- LONG
- INTEGER
- SHORT
- BYTE
- DOUBLE
- FLOAT
- DATE
- DECIMAL128
- OBJECT_ID

Places to check:
- ClassMetaData: Add TypeMirror to validDictionaryTypes
- RealmFieldType: Add new STRING_TO_X_MAP entry
- RealmFieldType: Add entry to "isValid"
- Constants.RealmFieldType: Add new STRING_TO_X_MAP entry
- Constants: Add new entry to DICTIONARY_ELEMENT_TYPE_TO_REALM_TYPES
- OsObjectBuilderTypeHelper:Add new entry to QUALIFIED_MAP_VALUES
- OsObjectBuilder: Add corresponding accessor methods, e.g. "addIntegerValueDictionary"
- OsObjectBuilder: Add a new item callback for the new type, e.g. "private static ItemCallback<Map.Entry<String, Integer>> integerMapItemCallback"
- OsObjectBuilder: Add a new native "add" method, e.g. "private static native void nativeAddIntegerDictionaryEntry(long dictionaryPtr, String key, int value)"
- io_realm_internal_objectstore_OsObjectBuilder.cpp: implement method mentioned above
- Property: Add a new case in convertFromRealmFieldType
- Property: Add a new case in convertToRealmFieldType
- RealmProxyClassGenerator: Add "Constants.RealmFieldType.STRING_TO_X_MAP" entry in "emitCreateExpectedObjectSchemaInfo"
- RealmDictionary: Add new branch to "getManager" (add new operator if needed, otherwise use "BoxableValueOperator")
- RealmProcessorTest: add corresponding field to "RealmDictionaryModel.java" class in assets

### Note: added new data types to other tests that require exhaustive evaluation on the type but the tests aren't ready yet.

### CODE REVIEW

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt [hunk 1]
```diff
@@ -296,6 +306,108 @@ class DictionaryTests {
...
+    fun managed_get() {
+        realm.executeTransaction { transactionRealm ->
+            transactionRealm.copyToRealm(initDictionaryClass(true))
+        }
+
+        val instance = realm.where<DictionaryClass>()
+                .findFirst()
+        assertNotNull(instance)
+
+        with(instance) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:352]: Would it somehow be possible to refactor these tests to use enum switches like our other Collection tests? There are so many supported datatypes that it is very easy to forget some, so using an enum makes sure we catch all of them, e.g. like https://github.com/realm/realm-java/blob/master/realm/realm-library/src/androidTest/java/io/realm/RealmResultsTests.java#L813

This is especially important later if we add a new datatype, as figuring out all the the tests that need to change is pretty hard.

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt [hunk 2]
```diff
@@ -513,47 +674,235 @@ class DictionaryTests {
...
+            assertNotNull(dictionaryFromRealm)
+
+            assertEquals(date1, dictionaryFromRealm[KEY_HELLO])
+            assertEquals(date2, dictionaryFromRealm[KEY_BYE])
+            assertNull(dictionaryFromRealm[KEY_NULL])
+        }
+    }
+
+    @Test
+    fun copyToRealm_decimal128() {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:884]: Missing UUID test? Same comment as before. Maybe convert all of these to private functions being called from a "super"-method that runs through an enum-switch?

File: realm/realm-library/src/androidTest/kotlin/io/realm/entities/DictionaryClass.kt [hunk 3]
```diff
@@ -17,26 +17,50 @@
...
 import java.util.*
 
 open class DictionaryClass : RealmObject() {
 
-//    var myEmbeddedObjectDictionary: RealmDictionary<EmbeddedSimpleChild>? = null
-
-    // TODO: ObjectId and other Java primitives
-    var myDogDictionary: RealmDictionary<MyDog>? = null
-    var myUUIDDictionary: RealmDictionary<UUID>? = null
+    var myRealmModelDictionary: RealmDictionary<MyRealmModel>? = null
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/kotlin/io/realm/entities/DictionaryClass.kt:28]: Do we allow this to be nullable? I thought we checked for this in the annotation processor. But that shouldn't be the case. Just like lists are not allowed to be nu null this way.

**Review by cmelchior — APPROVED** [review]
I must admit I got a bit lost in all the changes, but from what I can tell it looks fine. I left some suggestions for how we can improve the tests.

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt (hunk 1 above)
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:352]: Yes, I created a ticket for that. I will address that in one of the upcoming PRs.

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt (hunk 2 above)
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:884]: Sure, I will refactor it in the upcoming PR.

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt [hunk 4]
```diff
@@ -48,8 +50,16 @@ private const val VALUE_HELLO = true
...
 private const val VALUE_BYE_STRING = "BYE"
-private const val VALUE_HELLO_INTEGER = 42
-private const val VALUE_BYE_INTEGER = 666
+private const val VALUE_HELLO_NUMERIC = 42
+private const val VALUE_BYE_NUMERIC = 666
+private val VALUE_HELLO_DATE = Date()
+private val VALUE_HELLO_OBJECT_ID = ObjectId()
+private val VALUE_HELLO_UUID = UUID.randomUUID()
+private val VALUE_HELLO_DECIMAL128 = Decimal128(VALUE_HELLO_NUMERIC.toLong())
+private val VALUE_HELLO_BYTE_ARRAY = ByteArray(2).apply {
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:59]: It appears Kotlin doesn't allow calling `set` with a null value
Screenshot 2021-02-16 at 10 23 10

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt (hunk 2 above)
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:884]: And no, the UUID test was already there from the previous PR - that's why it doesn't appear in the diff.

File: realm/realm-library/src/androidTest/kotlin/io/realm/entities/DictionaryClass.kt (hunk 3 above)
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/entities/DictionaryClass.kt:28]: Created https://github.com/realm/realm-java/issues/7311 to handle it.

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt:59]: There is no support currently for nullable bytes in boxed byte arrays the SDK. I created an issue for this: https://github.com/realm/realm-java/issues/7312

File: realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt [hunk 5]
```diff
@@ -296,6 +306,108 @@ class DictionaryTests {
...
+                myLongDictionary!!.putAndAssert(KEY_HELLO, VALUE_HELLO_NUMERIC.toLong())
+                myShortDictionary!!.putAndAssert(KEY_HELLO, VALUE_HELLO_NUMERIC.toShort())
+                myDoubleDictionary!!.putAndAssert(KEY_HELLO, VALUE_HELLO_NUMERIC.toDouble())
+                myByteDictionary!!.putAndAssert(KEY_HELLO, VALUE_HELLO_NUMERIC.toByte())
+                myDateDictionary!!.putAndAssert(KEY_HELLO, Date())
+                myObjectIdDictionary!!.putAndAssert(KEY_HELLO, ObjectId())
+                myUUIDDictionary!!.putAndAssert(KEY_HELLO, UUID.randomUUID())
+                myDecimal128Dictionary!!.putAndAssert(KEY_HELLO, Decimal128(42))
+
+                val previousValue = myByteArrayDictionary!!.put(KEY_HELLO, VALUE_HELLO_BYTE_ARRAY)
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/androidTest/kotlin/io/realm/DictionaryTests.kt]: Same as above.

**Pull request merged**